include/crpropa/PhotonBackground.h

```diff
@@ -2,30 +2,50 @@
 #define CRPROPA_PHOTONBACKGROUND_H
 
 #include "crpropa/Common.h"
+#include "crpropa/Referenced.h"
 
 #include <string>
 
 namespace crpropa {
-
 /**
- * \addtogroup EnergyLosses
+ * \addtogroup PhotonFields
  * @{
  */
 
-
 /**
  @class PhotonField
- @brief photon field class fully defined with a range of photon energies, redshift and the field's density
+ @brief Abstract base class for photon fields.
  */
-class PhotonField {
+class PhotonField: public Referenced {
 public:
-	PhotonField(std::string fieldName, bool hasRedshiftDependence = true);
-	PhotonField() : PhotonField("CMB", false) {};
+	virtual double getPhotonDensity(double ePhoton, double z = 0.) const {
```

Comment:
In this case all constructors in the interaction modules .h files require explicitly if they are supposed to expect a ```TabularPhotonField``` or a ```BlackbodyPhotonField```. Having a constructor for each type of photon field is possible but would also make it more difficult to introduce more types of photon fields in the future. May I not be seeing something simpler here?

Reply:
This of course requires that the interaction classes have only pointers (ref_ptr) to the PhotonField in their constructor and not a copy of the field as now. 
A class can still be accessed via a pointer to a parent, only no object of type A can be constructed if there are only pure virtual methods. 
ref_ptr<A> acts as pointer A* , thus the confusion? Making them pure virtual only prevents that an abstract PhotonField is constructed.  See e.g. the RestrictToRegion module that has a pointer to a Surface which has pure virtual methods.
However, as this is similar to the MagneticField to have the no field defined as default, this can also stay.

Reply:
Good point indeed. Current implementation follows your idea and uses ref_ptr to PhotonField in all interaction modules while PhotonField now is an abstract class. Hate to bug you with it but in the current implementation it seems that in the steering file a call like
```
CMB = BlackbodyPhotonField("CMB", 2.73)
PPP = PhotoPionProduction(CMB)
```
returns a type error, explicitly demanding a ref_ptr<PhotonField> object (obviously...):
```
NotImplementedError: Wrong number or type of arguments for overloaded function 'new_PhotoPionProduction'.
  Possible C/C++ prototypes are:
    crpropa::PhotoPionProduction::PhotoPionProduction(crpropa::ref_ptr<rpropa::PhotonField>, ...)
```
But why does this occur in the Python call? The syntax is very similar to how magnetic fields are passed to the propagation modules and there we do not see this behaviour. Do you have an educated guess here?

Reply:
> You also have to make some changes in the 2_headers.i file. For all other ref_ptrs there are some declaration like this
> 
> `%ignore operator crpropa::MagneticField*;`
> 
> and this
> `%implicitconv crpropa::ref_ptr<crpropa::MagneticField>; %template(MagneticFieldRefPtr) crpropa::ref_ptr<crpropa::MagneticField>; %include "crpropa/magneticField/MagneticField.h"`
> 
> in the case of the MagneticField.
That totally did the trick, thanks a huge lot!!!

Reply:
solved by @lukasmerten, trick was to update the python 2_headers.i file accordingly